### PR TITLE
fix(core): minor audit improvements

### PR DESCRIPTION
**Pull request by sdbondi**
Description
---
fix(core): make max_input_maturity fallible
add details on the usage of core hash domains
include details on all fallible transaction input methods
minor comment update in coinbase_builder
fix(core/aggregate_body): remove unused buggy function min_input_maturity
tests(core/aggregate_body_validator): add test for unsorted aggregate body and comment on new_sorted_unchecked call
chore(core/transaction_builder): remove unnecessary mutation for more ergonomic api
chore(core): improve clarity of check_coinbase_output
fix(core): remove unused git link for monero

Motivation and Context
---
max_input_maturity: should be fallible rather than possibly return the incorrect result.
min_input_maturity: returns an incorrect result (u64::MAX) if there are no inputs. Removed because it was unused.
check_coinbase_output: had unwraps and a comment indicating something _should_ happen to prevent the panic. For clarity, updated the code so that it can easily be seen as correct.
transaction_builder: add components with a mutable reference, possibly unclear how the references are mutated. More ergonomic and useful to pass in an iterator.
monero dependency was using a git link and `version` ambiguously. Cargo seemed to use the crate and ignore the git field. 
Added an explicit test for validation failure on unordered tx components in an aggregate body.

How Has This Been Tested?
---
Existing tests and new unit test

What process can a PR reviewer use to test or verify this change?
---

Breaking Changes
---

- [x] None
- [ ] Requires data directory on base node to be deleted
- [ ] Requires hard fork
- [ ] Other - Please specify

### CODE REVIEW

**Comment by sdbondi** [comment]
This PR is broken down into commits. Each one can be reviewed independently. Happy to revert any one of them as necessary.

File: base_layer/core/src/transactions/transaction_components/transaction.rs [hunk 1]
```diff
@@ -97,19 +92,19 @@ impl Transaction {
...
-    pub fn min_spendable_height(&self) -> u64 {
+    pub fn min_spendable_height(&self) -> Result<u64, TransactionError> {
         self.body.min_spendable_height()
     }
 
-    /// This function adds two transactions together. It does not do cut-through. Calling Tx1 + Tx2 will result in
-    /// vut-through being applied.
+    /// This function adds two transactions together. It does not do cut-through but simply sums up the offsets and
+    /// extends all inputs, outputs and kernels.
     pub fn add_no_cut_through(mut self, other: Self) -> Self {
```
Comment (SWvheerden) [inline base_layer/core/src/transactions/transaction_components/transaction.rs]:
```suggestion
    pub fn add(mut self, other: Self) -> Self {
```
I dont see why we need to reference cut-through here as Tari does not use or support the use of cut-through anymore

Comment (SWvheerden) [reply base_layer/core/src/transactions/transaction_components/transaction.rs]: perhaps best to just delete this and use the impl add

File: base_layer/core/src/transactions/transaction_protocol/sender.rs [hunk 2]
```diff
@@ -72,38 +72,49 @@ pub(crate) struct OutputPair {
...
+    /// The TransactionOutput received from the recipient.
     pub recipient_output: Option<TransactionOutput>,
+    /// The partial kernel excess received from the recipient.
     pub recipient_partial_kernel_excess: PublicKey,
+    /// The partial kernel signature received from the recipient.
     pub recipient_partial_kernel_signature: Signature,
+    /// The partial kernel offset received from the recipient.
     pub recipient_partial_kernel_offset: PrivateKey,
-    // The sender's portion of the public commitment nonce
+    /// The change output details. This may be None if no change is required.
```
Comment (SWvheerden) [inline base_layer/core/src/transactions/transaction_protocol/sender.rs:87]:
```suggestion
    /// The change output details. Always required
```

**Review by SWvheerden — COMMENTED** [review]
Looks good, one nit, and one wrong comment

File: base_layer/core/src/transactions/transaction_protocol/sender.rs (hunk 2 above)
Comment (sdbondi) [reply base_layer/core/src/transactions/transaction_protocol/sender.rs:87]: In that case, I'll have to see how we handle none

Comment (SWvheerden) [reply base_layer/core/src/transactions/transaction_protocol/sender.rs:87]: On build it will throw an error

Comment (sdbondi) [reply base_layer/core/src/transactions/transaction_protocol/sender.rs:87]: That doesnt appear to be true. See `try_build_transaction` and `validate`

Comment (SWvheerden) [reply base_layer/core/src/transactions/transaction_protocol/sender.rs:87]: You are correct, dont know why I got that idea in my head.

File: base_layer/core/src/transactions/transaction_components/wallet_output.rs [hunk 3]
```diff
@@ -247,7 +247,7 @@ impl WalletOutput {
         ))
     }
 
-    pub async fn as_transaction_output<KM: TransactionKeyManagerInterface>(
+    pub async fn to_transaction_output<KM: TransactionKeyManagerInterface>(
```
Comment (SWvheerden) [inline base_layer/core/src/transactions/transaction_components/wallet_output.rs:250]: I dont agree with this change. 
`to_T` implies it takes `self`
`as_T` implies it takes `&self`

Comment (sdbondi) [reply base_layer/core/src/transactions/transaction_components/wallet_output.rs:250]: nope
https://rust-lang.github.io/api-guidelines/naming.html#ad-hoc-conversions-follow-as_-to_-into_-conventions-c-conv

Comment (sdbondi) [reply base_layer/core/src/transactions/transaction_components/wallet_output.rs:250]: `into_` - takes self
`to_` - usually `&self` some greater cost to the conversion
`as_` - usually `&self` cheap/basically free conversion

Comment (SWvheerden) [reply base_layer/core/src/transactions/transaction_components/wallet_output.rs:250]: Thanks, I stand corrected.

**Review by SWvheerden — APPROVED** [review]

File: base_layer/core/src/transactions/transaction_components/wallet_output.rs (hunk 3 above)
Comment (AaronFeickert) [reply base_layer/core/src/transactions/transaction_components/wallet_output.rs:250]: TIL!

**Pull request merged**